store/view_utilities.py

```diff
@@ -0,0 +1,50 @@
+from collections import defaultdict
+from store.models import Order, OrderItem, Product, Store
+
+
+def get_products_and_quantities_from_bag(request):
+    products = []
+    if "bag" in request.session:
+        for product in request.session["bag"]:
```

Comment:
here you are retrieving each `Product` instance from the db in a loop which can be inefficient and result in n+1 queries. Consider using a single query to fetch all products at once.

Take a look at this code. It looks the same but the key difference is the amount of trips to the db.

```
def get_products_and_quantities_from_bag(request):
    products = []
    if "bag" in request.session:
        product_ids = [product["product_id"] for product in request.session["bag"]]
        
        # Fetch all product instances in one query
        product_instances = Product.objects.filter(id__in=product_ids)
        
        # Create a dictionary to map product IDs to product instances
        product_dict = {product.id: product for product in product_instances}
        
        for product in request.session["bag"]:
            product_instance = product_dict.get(product["product_id"])
            if product_instance:
                products.append(
                    {
                        "product": product_instance,
                        "quantity": product["quantity"],
                        "image": product_instance.image.url,
                    }
                )
    return products
```